integrated_channels/sap_success_factors/client.py

```diff
@@ -148,15 +148,12 @@ def create_course_completion(self, user_id, payload):
             # TODO:
             #   if/when we decide we should use the generic course completion path
             #   for all customers, we should update the _payload_data method on the
-            #   SapSuccessFactorsLearnerDataTransmissionAudit class instead of doing this json load/dump
-            # Explanation:
-            #  json.loads loads "true" as True and when dumped again will be `true` (without quotes)
-            #  The initial payload explicitly sets the value to "true" *before* the dump, and thus it dumps with quotes.
+            #   SapSuccessFactorsLearnerDataTransmissionAudit class instead of doing this
             payload_to_update = json.loads(payload)
-            payload = json.dumps(payload_to_update)
+            payload_to_update['courseCompleted'] = bool(payload_to_update['courseCompleted'] == 'true')
```

Comment:
So, I've also seen in the transmission audit records that we are sending `courseCompleted = false` in addition to `courseCompleted = true`.  In my testing with both our pseudo-tenant and the customer tenant I was able to send a completion event with a value of `false`.  I was able to then send a second completion event with value of `true`.  At that point the completion record became "locked" and I couldn't send any more completion events without triggering the "duplicate completions" response from the tenant. 

I guess my point is here that our system is not only sending `true`, but it's also sending `false` for the `courseCompleted` value, and it's not clear to me if this latest implementation only handles the true scenario or if it handles both.  Can you clarify?

Reply:
Quick overview of process (as of the changes in this pr, not the previous bad code I wrote):
1. The audit model forms the payload with the following: `courseCompleted="true" if self.course_completed else "false",` (ie it specifically makes boolean values into strings) (this is in the `_payload_data` method of the audit model)
2. The payload is exported as a dictionary (in the`_payload_data` method of the audit model)
3. Then json dumped which turns the whole dictionary into a string. (in the `serialize` method of the audit model)
4. In the client when using the new endpoint, we then "load" the payload (this translates a string back into a dictionary so we can access specific values easily)
5. change the string boolean values into actual booleans
4. dump the payload (which puts them into javascript-style booleans: ie `true` and `false`)

Example of this logic:
```
>>> dict = {"truevalue": "true", "falsevalue": "false"}  # This is what the payload looks like before the serialize method
>>> dump = json.dumps(dict) 
>>> dump
'{"truevalue": "true", "falsevalue": "false"}'  # This is what it looks like after the serialize method
>>> loaded = json.loads(dump)  # This is the first thing we do in this client logic
>>> loaded
{'truevalue': 'true', 'falsevalue': 'false'}
>>> loaded['truevalue'] = bool(loaded['truevalue'] == 'true')
>>> loaded
{'truevalue': True, 'falsevalue': 'false'}
>>> loaded['falsevalue'] = bool(loaded['falsevalue'] == 'true')
>>> loaded
{'truevalue': True, 'falsevalue': False}
>>> dumped_final = json.dumps(loaded)
>>> dumped_final
'{"truevalue": true, "falsevalue": false}'
```
(This might be confusing given I have `falsevalue == 'true'` in the logic there, but I wanted to show how each boolean appears in the final outcome, even when using the same condition as we are using in the actual code. 

The original change in this PR that the style guide decided was unnecessary which might be a little more obvious was:
`payload_to_update['courseCompleted'] = True if payload_to_update['courseCompleted'] == 'true' else False`

To specifically answer your question, both Python booleans are translated into their lowercase versions without quotes when dumped back into the string version of the payload.

What I did previously to this pr was simply dump the payload's value which was `"true"` without converting to a valid boolean, so it actually ended up in the payload again as `"true"`. Which makes sense. Load > dump > load > dump should end up with the same values throughout if making no other changes. I don't know where in my original testing it somehow magically became a valid boolean. I think this was due to the test conditions being wrong.

But, with an actual valid boolean in a json.dumps call it will be converted to JS style booleans `true/false` with no quotes.

Reply:
Awesome, thanks for the detailed explanation!  The dropped `else False` would have 100% alleviated my concern here -- I guess I don't see why it would be better style to not include it.  I mean I get that the computer will handle the false condition just fine, but it's the humans that read the code who appreciate these crutches...